### PR TITLE
LIB 274 - Handle Custom responses in Webhooks/Codeboxes

### PR DESCRIPTION
**Pull request by opalczynski**
Guys @dancio @Zhebr @23doors , before make some test and usage, could you please take a look and tell me if you like/dislike this approach? Would be very thankful :)

I have one more question here:
What content types we should support ad-hoc? 
I want to give lib user ability to define their own handlers, but think we should have some ready to use immediately. Currently only: application/json and text/plain are supported.

### CODE REVIEW

**Comment by 23doors** [comment]
Looks good in my opinion. Handler approach is pretty neat. I think that json and plan text are the only 2 content types that we need at start.

**Comment by dancio** [comment]
lgtm, cloud u fix PR name?

**Comment by opalczynski** [comment]
@dancio I can fix everything ;)

**Comment by opalczynski** [comment]
@dancio @23doors @Zhebr 

I have some issue now :)

Consider this:
CodeBox code:
print(12)

the webhook.run() will return:
{"status":"success","duration":9703,"result":{"stderr":"","stdout":"12"},"executed_at":"2015-12-08T16:21:19.429648Z","id":17}

But,
CodebBox code:
set_response(HttpResponse(status_code=200, content='{"raz": 1}', content_type='application/json'))

Then the webhook connected to this codebox will return:
{'raz': 1}

And this is good - we want to overwrite the standard behaviour completely. But in LIB I can't separate this two situations.

Both of returned types are dict, both possibly can have the same keys (user can define such fields as status, result and even the executed_at). For now - I left the for sure - not bulletproof solution :)

The second problem is that - I think it's rather bad to return sometimes the WebhookTrace object and sometimes a user defined data - any ideas?

**Comment by 23doors** [comment]
I guess there is no way around it and we need to sometimes return WebhookTrace object and sometimes user defined data. I guess that's the use case that custom response was after anyway.

File: syncano/models/custom_response.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    response_handler = CustomResponseHandler()
+
+    @property
+    def content(self):
+        return self.response_handler.process_response(self.result)
+
+    @property
+    def status_code(self):
+        if not self.result:
```
Comment (Zhebr) [inline syncano/models/custom_response.py]: I wonder, why do you prefer negative conditional? Instead you could write
```
if self.result:
    return self.result.get('response', {}).get('status')
```
I have nothing against - it's just interesting.

**Comment by Zhebr** [comment]
neat

File: syncano/models/custom_response.py (hunk 1 above)
Comment (forgems) [inline syncano/models/custom_response.py]: return self.result.get('status') if self.result else None
albo
return (self.result or {}).get('status')

Comment (forgems) [inline syncano/models/custom_response.py]: A tak ogólnie to zwarcanie None (NULL, nil) w miejscu gdzie się oczekuje kolekcji nie jest najlepszą praktyką.

Comment (opalczynski) [inline syncano/models/custom_response.py]: Well :) This is how JSONField is implemented in LIB and been there a long time ago before me :) 

```
    def to_python(self, value):
        if value is None:
            return

        if isinstance(value, six.string_types):
            value = json.loads(value)
        return value
```

When you create an object like this:
```
c = CodeBoxTrace()
```

And then try to get result - it's not populated yet;

It's the same story as django:

```
u = User()

print(u.username)  # -> u''

```
Except that in django in above case at least types match.

Comment (forgems) [inline syncano/models/custom_response.py]: Forgot about using english :)
I understand this. This was just a note, to keep this in mind in future. Keeping everything in JsonFields is very handy, but also very dangerous in terms of contract and interfaces.

Comment (opalczynski) [inline syncano/models/custom_response.py]: Agree :)

**Pull request merged**